reference/reference.go

```diff
@@ -0,0 +1,181 @@
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/docker/distribution/digest"
+	distreference "github.com/docker/distribution/reference"
+	"github.com/docker/docker/image/v1"
+)
+
+const (
+	// DefaultTag defines the default tag used when performing images related actions and no tag string is specified
+	DefaultTag = "latest"
+	// DefaultHostname is the default built-in hostname
+	DefaultHostname = "docker.io"
+	// LegacyDefaultHostname is automatically converted to DefaultHostname
+	LegacyDefaultHostname = "index.docker.io"
+	// DefaultRepoPrefix is the prefix used for defautl repositories in default host
+	DefaultRepoPrefix = "library/"
+)
+
+// Reference is an opaque object reference identifier that may include
+// modifiers such as a hostname, name, tag, and digest.
+type Reference interface {
+	// String returns the full reference
+	String() string
+}
+
+// Named is an object with a full name
+type Named interface {
+	Reference
+	Name() string
+}
+
+// NamedTagged is an object including a name and tag.
+type NamedTagged interface {
+	Named
+	Tag() string
+}
+
+// Canonical reference is an object with a fully unique
+// name including a name with hostname and digest
+type Canonical interface {
+	Named
+	Digest() digest.Digest
+}
+
+// ParseNamed parses s and returns a syntactically valid reference implementing
+// the Named interface. The reference must have a name, otherwise an error is
+// returned.
+// If an error was encountered it is returned, along with a nil Reference.
+func ParseNamed(s string) (Named, error) {
+	named, err := distreference.ParseNamed(s)
+	if err != nil {
+		return nil, fmt.Errorf("Error parsing reference: %q is not a valid repository/tag", s)
+	}
+	_, remoteName := distreference.SplitHostname(named)
+	if err := validateName(remoteName); err != nil {
+		return nil, err
+	}
+	// If both tag and digest, ignore tag
+	if _, isTagged := named.(NamedTagged); isTagged {
+		if canonical, isCanonical := named.(Canonical); isCanonical {
+			named, err = WithDigest(named, canonical.Digest())
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return Normalize(named), nil
+}
+
+// Normalize returns a reference in its normalized form, meaning repo name
+// will not contain default hostname nor library/ prefix for official images.
+func Normalize(ref Named) Named {
+	host, remoteName := SplitHostname(ref)
+	if host == DefaultHostname {
+		if strings.HasPrefix(remoteName.Name(), DefaultRepoPrefix) {
+			return withNewName(ref, strings.TrimPrefix(remoteName.Name(), DefaultRepoPrefix))
+		}
+		return remoteName
+	}
+	return ref
+}
+
+// WithHostname returns a referece with full hostname and index path.
+func WithHostname(ref Named) Named {
+	host, remoteName := SplitHostname(ref)
+	return withNewName(ref, host+"/"+remoteName.Name())
+}
+
+// SplitHostname splits a named reference into a
+// hostname and name string. If no valid hostname is
+// found, the hostname is empty and the full value
+// is returned as name
```

Comment:
This comment doesn't seem to be true. If there's no hostname, it returns the default index name.

This seems to be the main difference from `distreference.SplitHostname`, besides the extra validation, and it looks like the reason `ParseNamed` uses `distreference.SplitHostname` instead of this version.

I feel this function should be renamed to something else so it doesn't get confused with the other version, and to make it clearer it's doing more than just splitting on a hostname boundary.

Perhaps `IndexAndRemoteName`?